### PR TITLE
internal/v4: resolve ids

### PR DESCRIPTION
**Pull request by rogpeppe**
This allows all charm store paths to contain charm URLs that are
not fully specified.

### CODE REVIEW

File: internal/charmstore/store_test.go [hunk 1]
```diff
@@ -50,6 +50,94 @@ func (s *StoreSuite) TestAddCharm(c *gc.C) {
...
+	expect:  []string{"cs:precise/wordpress-23"},
+}, {
+	inStore: []string{"cs:precise/wordpress-23", "cs:trusty/wordpress-24", "cs:foo/bar-434"},
+	expand:  "wordpress",
+	expect:  []string{"cs:precise/wordpress-23", "cs:trusty/wordpress-24"},
+}, {
+	inStore: []string{"cs:precise/wordpress-23", "cs:trusty/wordpress-23", "cs:trusty/wordpress-24"},
+	expand:  "wordpress-23",
+	expect:  []string{"cs:precise/wordpress-23", "cs:trusty/wordpress-23"},
+}, {
```
Comment (urosj) [inline internal/charmstore/store_test.go:158]: Add a test with "nothing to expand" like:
inStore: []string{"cs:precise/wordpress-23", "cs:trusty/wordpress-24", "cs:foo/bar-434"}
expand: "precise/wordpress-23"
expect: []string{"cs:precise/wordpress-23"}

File: internal/charmstore/store_test.go [hunk 2]
```diff
@@ -50,6 +50,94 @@ func (s *StoreSuite) TestAddCharm(c *gc.C) {
 	c.Assert(err, jc.Satisfies, mgo.IsDup)
 }
 
+var expandURLTests = []struct {
+	inStore []string
+	expand  string
+	expect  []string
+}{{
+	inStore: []string{"cs:precise/wordpress-23"},
```
Comment (bac) [inline internal/charmstore/store_test.go:139]: An inStore entry that is not matched would enhance the test.  Something similar but rejected by the query.

File: internal/charmstore/store_test.go [hunk 3]
```diff
@@ -50,6 +50,94 @@ func (s *StoreSuite) TestAddCharm(c *gc.C) {
...
+	expect:  []string{"cs:precise/wordpress-23", "cs:trusty/wordpress-23"},
+}, {
+	inStore: []string{"cs:~user/precise/wordpress-23", "cs:~user/trusty/wordpress-23"},
+	expand:  "~user/precise/wordpress",
+	expect:  []string{"cs:~user/precise/wordpress-23"},
+}, {
+	inStore: []string{"cs:~user/precise/wordpress-23", "cs:~user/trusty/wordpress-23"},
+	expand:  "~user/wordpress",
+	expect:  []string{"cs:~user/precise/wordpress-23", "cs:~user/trusty/wordpress-23"},
+}}
```
Comment (frankban) [inline internal/charmstore/store_test.go:174]: Nice tests.

File: internal/v4/api.go [hunk 4]
```diff
@@ -57,13 +56,59 @@ func New(store *charmstore.Store) http.Handler {
 	return h
 }
 
-func (h *handler) resolveURL(url *charm.URL) error {
-	if url.Series == "" || url.Revision == -1 {
-		return fmt.Errorf("id resolving not implemented yet")
+// ResolveURL resolves the series and revision of the given URL
+// if unspecified.
+func ResolveURL(store *charmstore.Store, url *charm.URL) error {
```
Comment (bac) [inline internal/v4/api.go:63]: Using 'resolve' as the description to what ResolveURL does is circular. Would be nice to specify the expanded URL is returned in the url pointer, or unchanged.

**Comment by frankban** [comment]
:+1: thank you!

File: internal/charmstore/store.go [hunk 5]
```diff
@@ -43,6 +44,34 @@ func (s *Store) AddCharm(url *charm.URL, c charm.Charm) error {
...
+}
+
+func matchURL(url, pattern *charm.URL) bool {
+	if pattern.Series != "" && url.Series != pattern.Series {
+		return false
+	}
+	if pattern.Revision != -1 && url.Revision != pattern.Revision {
+		return false
+	}
+	return true
```
Comment (urosj) [inline internal/charmstore/store.go]: check the same name as well.

**Comment by bac** [comment]
:+1: , thanks

**Comment by urosj** [comment]
LGTM (with small changes in tests and proper matchURL return instead of return true)

File: internal/v4/api.go (hunk 4 above)
Comment (rogpeppe) [inline internal/v4/api.go:63]: done

File: internal/charmstore/store_test.go (hunk 2 above)
Comment (rogpeppe) [inline internal/charmstore/store_test.go:139]: Good point. Done.

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Build failed: Attempt to land pull request failed
        build url: http://ci.jujugui.org:8080/job/charmstore-merge/26

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**